### PR TITLE
BEAST_DEFINE_TESTSUITE_MANUAL(ProtocolJson,protocol,ripple);

### PR DESCRIPTION
**Pull request by sublimator**
Just leaving this quick hack here as a starting point for you C++ gurus to work off/consider. This is a manually run test that dumps the protocol definitions in json, which I used to update ripple-lib-java for omnigate.com. 

It's too error-prone to do manually so I hacked this in to spit out json, and then loaded the json inside a java unit test and made sure all the Java definitions (using native enums/classes etc) were in sync.

It would be quite a useful command/feature to have, to keep the various ripple-lib-* up to date. It might be more nicely exposed as an RPC command.

### CODE REVIEW

**Comment by MarkusTeufelberger** [comment]
This seems to be a sample output: https://github.com/sublimator/ripple-lib-java/blob/d55fe4336cd8a190f41ffe6911006d75b7bf26fd/ripple-core/src/test/resources/protocol.json

**Comment by sublimator** [comment]
Yeah, the schema/code could be improved but I think this would be a very useful feature

File: src/ripple/protocol/impl/SField.cpp [hunk 1]
```diff
@@ -401,4 +401,17 @@ SField::getField (std::string const& fieldName)
     return sfInvalid;
 }
 
+Json::Value SField::allFieldsJson ()
+{
+    Json::Value all(Json::arrayValue);
+    for (auto const& pair : knownCodeToField)
+    {
+        if (pair.second->isBinary())
```
Comment (sublimator) [inline src/ripple/protocol/impl/SField.cpp]: This kind of makes the 'isBinary' output pointless ... 

The reason it was there in the first place was that that there's some Hash256 fields like `index` and `hash` that are useful to declare, such that any Json -> dehydrated class works simply.  There's also other Amount fields like "taker_gets_funded" etc in offers json.

**Comment by nbougalis** [comment]
I don't like the commenting out of several lovely checks here. While I am supportive of this, I think it would probably be better if it was a standalone tool that parsed the source code. :thinking:

**Comment by sublimator** [comment]
Parse the C++ code with clang/regexes or sth? That would be a pain in the
ass to dev/maintain. Or you have sth else in mind? If you could do it in a
separate tool rather than abusing the unittest machinery that would be
cool. Were you thinking something like the manifest signing tool that
imports the xrpl/core (or whatever) lib?

Yeah, I don’t like commenting out checks, but I just kept it rebased so I
could spit out some JSON now and then. It’s been quite useful. Some people
asked me to do a bit of updating on ripple-lib-java a while back, and I was
getting a headache eyeballing all the changes  so I added some tests to
check the JSON matches all the static Java declarations.

**Comment by sublimator** [comment]
> parsed the source code.

Aaaaah. I see.

**Comment by sublimator** [comment]
Actually, now that I remember, (iirc) there are some things that were only
avail (easily) at runtime. I first messed with something like this using
gdb and a python plugin back in the day. Looked into clang api and was all
ugh, no time for this, then used test suite.

**Comment by nbougalis** [comment]
I would like to merge this but the commenting of several `static_assert` calls is a concern. Moreover, I'm pretty sure that this functionality should be part of a standalone tool that consumes a definition file and produces the requires files in the programming language of choice.

For example:

    generate-xrp-fields.py --lang=c++ fields.xrpl

And:

    generate-xrp-fields.py --lang=java fields.xrpl

**Comment by sublimator** [comment]
Right, moving the source of truth to something other than hard to parse C++
sounds good to me :)

This was just a pragmatic solution at the time.

Not offended if y’all close the PR

**Comment by sublimator** [comment]
@nbougalis 
Note, that it's not just the fields. It's also the formats (object schemas) and engine results. Maybe some other enums.

**Comment by sublimator** [comment]
https://raw.githubusercontent.com/sublimator/ripple-lib-java/master/ripple-core/src/test/resources/protocol.json
```
> Object.keys(require('./protocol'))

[ 'engineResults', 'fields', 'ledgerEntries', 'transactions' ]
```
